internal/pkg/agent/protection/policy.go

```diff
@@ -0,0 +1,196 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package protection
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+var (
+	ErrNotFound = errors.New("not found")
+)
+
+type AgentProtection struct {
+}
+
+// ValidatePolicySignature validates policy document signature, overlays with signed configuration
+// returns the policy with signed data overlaying the original policy, the matching signatureValidationKey
+// The signatureValidationKey parameter can be empty, in that case try to read the key from the policy.protected.signing_key
+// and validate the signature with that key.
+// This would guarantee that the signature validation key returned from the function is the valid key as well.
+//
+// There could be few cases this function needs to handle:
+// 1. The old policy format that doesn't have signature or protected data. Policy should returned without modifications or errors.
+// 2. New policy where the signatureValidationKey is nil. No previously known signature validation key, try to read one from the policy, validate and return.
+// 3. New policy format with the signatureValidationKey. Validate the signed data normally.
+func ValidatePolicySignature(log *logger.Logger, policy map[string]interface{}, signatureValidationKey []byte) (map[string]interface{}, []byte, error) {
+	var err error
+
+	const logPrefix = "Validate policy signature:"
+
+	log.Infof("%v passed signature validation key length: %v", logPrefix, len(signatureValidationKey))
+
+	// Try to get the signature validation key from the policy itself if there is no previously known signature validation key
+	if len(signatureValidationKey) == 0 {
+		signatureValidationKey, err = getPolicySignatureValidationKey(policy)
+		log.Infof("%v policy signature validation key length: %v, err: %v", logPrefix, len(signatureValidationKey), err)
+		if err != nil && !errors.Is(err, ErrNotFound) {
+			return nil, nil, err
+		}
```

Comment:
we could handle new len==0 here 
```golang
if len(signatureValidationKey) == 0 {
   log.Infof("%v signature validation key is not present, continue", logPrefix)
   return policy, signatureValidationKey, nil
}
``` 

and get rid of if nesting below

Reply:
flattened nested blocks with early returns